### PR TITLE
Remove duplicate Queries and duplicate Clicked Documents SFINT-2464

### PR DESCRIPTION
**Pull request by jeremierobert-coveo**
Remove duplicate Queries and duplicate Clicked Documents.
Also preserve order

### CODE REVIEW

File: src/components/UserActions/QueryList.ts [hunk 1]
```diff
@@ -88,7 +88,7 @@ export class QueryList extends Component {
      * @param bindings Bindings of the Search-UI environment.
      */
     constructor(public element: HTMLElement, public options: IQueryListOptions, public bindings: IComponentBindings) {
-        super(element, QueryList.ID, bindings);
+        super(element, QueryList.ID, bindings) /* istanbul ignore next Istanbul issue with next */;
```
Comment (tedre191) [inline src/components/UserActions/QueryList.ts]: why is istanbul not happy here ?

File: src/components/UserActions/ClickedDocumentList.ts [hunk 2]
```diff
@@ -98,6 +98,7 @@ export class ClickedDocumentList extends Component {
                 .filter(action => action.document && action.type === UserActionType.Click)
                 .sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime())
                 .reverse()
+                .reduce((acc, action) => (!acc.find(existing => existing.raw.uri_hash === action.raw.uri_hash) ? [...acc, action] : acc), [])
```
Comment (tedre191) [inline src/components/UserActions/ClickedDocumentList.ts]: I would add a comment and even split that into a seperate function
ex. `.reduce(filterDuplicates)` or something

File: src/components/UserActions/QueryList.ts (hunk 1 above)
Comment (jeremierobert-coveo) [reply src/components/UserActions/QueryList.ts]: istanbul can't digest supper call.

**Comment by louis-bompart** [comment]
PR should be well presented (i.e. a fitting name and a description if needed)

File: tests/components/UserActions/ClickedDocumentList.spec.ts [hunk 3]
```diff
@@ -9,6 +9,7 @@ import { UserActionType } from '../../../src/rest/UserProfilingEndpoint';
 describe('ClickedDocumentList', () => {
     const TEST_CLICKS = generate(20, i => {
         const document = Fake.createFakeResult();
+        document.uri = `document${Math.random()}`;
```
Comment (louis-bompart) [inline tests/components/UserActions/ClickedDocumentList.spec.ts]: We talked about it in a former review with @erocheleau: what's the point of Random in tests. That's, imho, undesirable and useless complexity.

File: tests/components/UserActions/ClickedDocumentList.spec.ts [hunk 4]
```diff
@@ -117,6 +118,54 @@ describe('ClickedDocumentList', () => {
         });
     });
 
+    it('should not show the same query twice', () => {
```
Comment (louis-bompart) [inline tests/components/UserActions/ClickedDocumentList.spec.ts]: This test is quite long, maybe put comment to indicate where are the phases of the tests and explain what you're doing.

File: tests/components/UserActions/ClickedDocumentList.spec.ts [hunk 5]
```diff
@@ -117,6 +118,54 @@ describe('ClickedDocumentList', () => {
...
+            if (button) {
+                button.click();
+            }
+
+            const list = mock.env.element.querySelector<HTMLOListElement>('.coveo-list');
+            expect(list.childElementCount).toBe(new Set(TEST_DUPLICATE.map(action => action.raw.uri_hash)).size);
+
+            // Check that the order is respected.
+            const sortedActions = TEST_DUPLICATE.reverse().reduce(
+                (acc, action) => (!acc.find(existing => existing.raw.uri_hash === action.raw.uri_hash) ? [...acc, action] : acc),
```
Comment (louis-bompart) [inline tests/components/UserActions/ClickedDocumentList.spec.ts]: No, you're reproducing the logic of the code, thus the tests does not really test per se.
Hardcode the input and output test values and compare the actual output to the test output.

File: tests/components/UserActions/QueryList.spec.ts [hunk 6]
```diff
@@ -110,6 +112,42 @@ describe('QueryList', () => {
...
+            const button = mock.cmp.element.querySelector<HTMLElement>('.coveo-more-less');
+            if (button) {
+                button.click();
+            }
+
+            const list = mock.env.element.querySelector<HTMLOListElement>('.coveo-list');
+            expect(list.childElementCount).toBe(new Set(TEST_DUPLICATE.map(action => action.query)).size);
+
+            // Check that the order is respected.
+            TEST_DUPLICATE.reverse()
```
Comment (louis-bompart) [inline tests/components/UserActions/QueryList.spec.ts]: No, you're reproducing the logic of the code, thus the tests does not really test per se.
Hardcode the input and output test values and compare the actual output to the test output.

File: tests/components/UserActions/ClickedDocumentList.spec.ts (hunk 3 above)
Comment (jeremierobert-coveo) [reply tests/components/UserActions/ClickedDocumentList.spec.ts]: The random part prevent depending on a specific string value.

File: tests/components/UserActions/ClickedDocumentList.spec.ts (hunk 5 above)
Comment (jeremierobert-coveo) [reply tests/components/UserActions/ClickedDocumentList.spec.ts]: hardcode ark, but why not 😐

File: tests/components/UserActions/QueryList.spec.ts (hunk 6 above)
Comment (jeremierobert-coveo) [reply tests/components/UserActions/QueryList.spec.ts]: ditto

File: tests/components/UserActions/ClickedDocumentList.spec.ts (hunk 3 above)
Comment (louis-bompart) [reply tests/components/UserActions/ClickedDocumentList.spec.ts]: Personally I would rather have documentTestUri that a random, that's some over-engineering imho.
What do you think @tedre191, @erocheleau ?

File: tests/components/UserActions/ClickedDocumentList.spec.ts (hunk 5 above)
Comment (louis-bompart) [reply tests/components/UserActions/ClickedDocumentList.spec.ts]: Well, let's think about what the value of a unit test. It should ensure that the 'black box' that is your code, when given a specific input give you back an expected output. If you repeat the black box in the tests, you're not unit-testing.

File: tests/components/UserActions/QueryList.spec.ts (hunk 6 above)
Comment (louis-bompart) [reply tests/components/UserActions/QueryList.spec.ts]: ditto ditto? dittoto ? dittotto ?

File: tests/components/UserActions/ClickedDocumentList.spec.ts (hunk 5 above)
Comment (jeremierobert-coveo) [reply tests/components/UserActions/ClickedDocumentList.spec.ts]: It's not reproduce. It just happen that the reduction to sort thing out is the same.

**Pull request merged**